dali/pipeline/data/tensor_vector.cc

```diff
@@ -172,240 +206,299 @@ void CopyImpl(DstBatch<DstBackend> &dst, const SrcBatch<SrcBackend> &src, const
-  DALI_ENFORCE(type() == src.type(),
-               make_string("Sample must have the same type as a target batch, current: ", type(),
-                           " new: ", src.type(), " for ", sample_idx, " <- ", src_sample_idx, "."));
-  DALI_ENFORCE(sample_dim() == src.shape().sample_dim(),
-               make_string("Sample must have the same dimensionality as a target batch, current: ",
-                           sample_dim(), " new: ", src.shape().sample_dim(), " for ", sample_idx,
-                           " <- ", src_sample_idx, "."));
-  DALI_ENFORCE(
-      GetLayout() == src.GetLayout(),
-      make_string("Sample must have the same layout as a target batch current: ", GetLayout(),
-                  " new: ", src.GetLayout(), " for ", sample_idx, " <- ", src_sample_idx, "."));
+  VerifySampleCopyConformance(src.type(), src.shape().sample_dim(), src.GetLayout(),
+                              shape()[sample_idx], src.shape()[src_sample_idx],
+                              make_string(" for ", sample_idx, " <- ", src_sample_idx, "."));
+
+  shape_.set_tensor_shape(sample_idx, src.shape()[src_sample_idx]);
+  tensors_[sample_idx].Copy(src.tensors_[src_sample_idx], order);
+  if (src.GetLayout().empty() && !GetLayout().empty()) {
+    tensors_[sample_idx].SetLayout(GetLayout());
+  }
+}
 
-  // Either the shape matches and we can copy data as is or the target is just an individual sample
-  bool can_copy = tensors_[sample_idx]->shape() == src.tensors_[src_sample_idx]->shape() ||
-                  (!tl_->has_data() && state_ == State::noncontiguous);
 
-  DALI_ENFORCE(
-      can_copy,
-      "Copying samples into TensorVector can happen either for exact shape match or when the "
-      "TensorVector is truly non contiguous. Either Resize first to the desired shape or reset the "
-      "TensorVector and SetSize for desired number of samples in non-contiguous mode.");
-
-  tensors_[sample_idx]->Copy(*src.tensors_[src_sample_idx], order);
+template <typename Backend>
+void TensorVector<Backend>::UnsafeCopySample(int sample_idx, const Tensor<Backend> &src,
+                                             AccessOrder order) {
+  // Bounds check
+  assert(sample_idx >= 0 && sample_idx < curr_num_tensors_);
+  VerifySampleCopyConformance(src.type(), src.shape().sample_dim(), src.GetLayout(),
+                              shape()[sample_idx], src.shape(),
+                              make_string(" for ", sample_idx, "."));
+
+  shape_.set_tensor_shape(sample_idx, src.shape());
+  tensors_[sample_idx].Copy(src, order);
+  if (src.GetLayout().empty() && !GetLayout().empty()) {
+    tensors_[sample_idx].SetLayout(GetLayout());
+  }
 }
 
+
 template <typename Backend>
 void TensorVector<Backend>::set_sample_dim(int sample_dim) {
   DALI_ENFORCE(
       !has_data(),
       "Setting sample dim is not allowed when batch is already allocated, use Resize instead.");
   sample_dim_ = sample_dim;
+  shape_.resize(shape_.num_samples(), sample_dim);
```

Comment:
Should we have set_sample_dim for shapes also? `resize()` pretty often used like that

Reply:
I will look into a separate PR with that